### ISSUE TITLE
[plugin-api-docs] Support OpenAPI 3.1

### ISSUE THREAD

**Issue by johanneswuerbach** [backstage/backstage#8090]
## Feature Suggestion

Support rendering `openapi: 3.1.x` API definitions in Backstage. OpenAPI 3.1.0 [was released](https://github.com/OAI/OpenAPI-Specification/releases/tag/3.1.0) as stable in the beginning of the year. 

## Possible Implementation

This likely depends on swagger-ui supporting OpenAPI 3.1, which it doesn't yet https://github.com/swagger-api/swagger-ui/issues/5891, but it might be a good preparation to update to swagger-ui v4 already as only [v4 will get OpenAPI 3.1 support]( https://github.com/swagger-api/swagger-ui/issues/5891#issuecomment-899512074).

## Context

Current backstage view when looking at an OpenAPI 3.1 schema:

![image](https://user-images.githubusercontent.com/864578/142193829-de4ff0a6-3e32-4202-95a7-265ab0447e8d.png)

**Comment by freben** [comment]
Makes sense! Is this something you might be able to work on?

**Comment by freben** [comment]
I see that we depend on `"swagger-ui-react": "^4.0.0-rc.3"` in the API plugin. Latest version of that as of this time is 4.1.0.

**Comment by Rugvip** [comment]
Yep, would make sense to update to the latest stable release. Opening that up to contributions in case anyone wants to beat us to it :grin:

**Comment by floriank** [comment]
So,

Hey  - since it'd be rather important to me (and the company) to get some movement on the OAS 3.1 front going, I've spent some time looking into this. The news is a bit bleak though:

1. Backstage depends on `swagger-ui-react` (see [here](https://github.com/backstage/backstage/blob/a5684d0d53d535360e14ca07e34d8a8423db7348/plugins/api-docs/package.json#L53-L53)
2. `swagger-ui-react` currently does _not_ support OAS 3.1, hence we get no API explorer for specs using that scheme version
3. The swagger authors decided to work on a new version of their editor first, which [will resolve the issue](https://github.com/swagger-api/swagger-ui/issues/5891#issuecomment-1172274595)

This unfortunately means either one of two things:

1. Sit and wait.
2. Replace the dependency with something like [OpenAPI Explorer](https://www.npmjs.com/package/openapi-explorer). I've no opinion on this package myself (and I am not associated).

Anyway, this issue therefore should probably no be a good first contribution :wink:.

Cheers :heart:

**Comment by floriank** [comment]
(I also should've realized that @johanneswuerbach mentioned most of this in the original issue description, mea culpa :facepalm:)

**Comment by floriank** [comment]
In an effort to provide a tiny bit more substance, here is a branch where I [switched out `swagger-ui-react` with `openapi-explorer`](https://github.com/floriank/backstage/tree/feat/openapi-explorer).

The result needs styling, obviously, but works rather nice, even with OAS 3.1:

![image](https://user-images.githubusercontent.com/498241/186738167-5f8f4e84-063c-4789-96d0-3d50383d1ed2.png)

**Comment by johanneswuerbach** [comment]
Still relevant 👍

**Comment by EduSemensati** [comment]
Is anyone looking into this?

**Comment by floriank** [comment]
not actively, as I don't really have time myself. Switching out the OpenAPI Explorer is easy, but I am no CSS magician, so it'd look like :hankey:.

**Comment by jetslambda** [comment]
Personal opinion: Tests in Backstage are intended for React components making openapi-explorer hard to test since it is a 'Lit' component (Google.) I was able to integrate it and make it look good but I was unable to fiddle with it enough to get all tests passing. Swagger.io intends to have OpenAPI3.1 compatibility in swagger-ui this Quarter. I think we should wait a few weeks and implement the upcoming version of swagger-ui.

https://swagger.io/blog/swagger-support-for-openapi-3-0-and-openapi-3-1/
```

Swagger UI and Swagger Client 
Rendering for OpenAPI 3.1 will soon follow as part of our [Swagger UI](https://github.com/swagger-api/swagger-ui) enhancements. In a nutshell, the ability to process and render OpenAPI 3.1 definitions/documents within Swagger UI will be enabled by the integration of [ApiDOM](https://github.com/swagger-api/apidom) with [Swagger Client](https://github.com/swagger-api/swagger-js). 

We’re expecting to have this completed in Q1 2023 which will bring the rendering capability also into our Swagger Editor tool.
```

**Comment by jetslambda** [comment]
My implementation is here if anyone would like to do something with it:

https://github.com/backstage/backstage/blob/88443dca751eacf7ba49c64a753908afd5c8fd7c/plugins/api-docs/src/components/OpenApiDefinitionWidget/OpenApiDefinition.tsx#L27

**Comment by freakyfelt** [comment]
nudging to appease the stalebot

**Comment by pjungermann** [comment]
https://github.com/swagger-api/swagger-ui/labels/version%3A%205.x
and https://github.com/swagger-api/swagger-ui/issues/8684 to watch for beta

**Comment by pjungermann** [comment]
FYI: swagger-ui v5.0.0 was released 2h ago.

**Comment by pjungermann** [comment]
In case you didn't see it yet, the linked PR will resolve this issue. It contains two 3.1.0 examples as well. A release of the type package was not necessary, they are still compatible.

If reviews go through, it should be part of next week's main release.
